src/docs/arch-prefetch.dox

```diff
@@ -0,0 +1,106 @@
+
+The \c available setting determines whether the pre-fetching functionality can be
+enabled for the given database, and starts up the pool of pre-fetching threads if
+it is turned on. The \c default setting determines whether pre-fetching is turned
+on or off for all sessions belonging to the connection.
+
+@subsection prefetch_api_session Session-Level Configuration
+\c WT_CONNECTION.open_session("prefetch=(enabled=true)")
+
+The session-level configuration can be used to override the connection-level
+configuration for specific sessions if desired.
+
+@section prefetch_other Interaction With Other Components
+
+@subsection prefetch_other_cache_busyness Considering current cache busyness
+Cache busyness increases in proportion to the eviction workload and the rate of pages
+being read into the cache. It is important to consider how pre-fetch impacts cache
+utilization as it may leave the cache in a state that increases operational latency.
+Pre-fetch stops adding page references to the queue as soon as it detects that we are
+close to hitting the eviction clean trigger. It will also skip processing page
+references that are already on the queue until space in the cache clears up.
+
+@subsection prefetch_other_refs Ensuring references on the pre-fetch queue remain valid
+Page references on the pre-fetch queue may become invalid if other operations
+that modify page references are happening in parallel. Eviction is not allowed to
+operate on any page references that are on the pre-fetch queue, because it may
+split pages and free the underlying ref/s. Since eviction has no way to coordinate
+with the queue to indicate references are no longer valid and should be removed, we
+take a conservative approach by preventing eviction on these refs. This condition is
+checked with the \c WT_REF_FLAG_PREFETCH flag. On top of this, the pre-fetch flag
+also guarantees that the corresponding parent page cannot be evicted until pre-fetch
+has processed the leaf page.
+
+@subsection prefetch_other_concurrency Concurrency
+Pre-fetch is a highly concurrent module and we rely on a few mechanisms to ensure
+that it runs smoothly with the rest of the system.
+- A connection level lock \c prefetch_lock is used for the pre-fetch queue, and is
+needed when adding and removing \c WT_PREFETCH_QUEUE_ENTRY items from it.
+- Whenever we modify a \c ref structure, e.g. when adding it to the pre-fetch queue
+and setting the corresponding \c WT_REF_FLAG_PREFETCH flag, it must be in the
+\c WT_REF_LOCKED state.
+- There are scenarios which fall outside of the typical eviction flow, where pages
+can be forcibly removed from the cache if eviction has exclusive eviction access to
+that file (see \c __wt_evict_file_exclusive_on). When this occurs, we need to remove
+all page references associated with that file's tree from the pre-fetch queue because
+the system may evict internal pages which still have child pages present on the queue.
```

Comment:
Ah, I see now: Eviction should not evict any internal pages having leafs in memory. If Prefetch is going to load a leaf page, it should instruct Eviction to leave alone all parent internal pages.
Is it where the overlap can happen?
But it was mentioned that Prefetch only operates on leaf pages.
I think, I could work it out but this moment puzzles me a little bit.
Would it be worth describing this interaction in more details in this doc?

Reply:
Yes, I wasn't sure whether I should even include this section in the first place. The pre-fetch doc is quite long already and I was thinking of cutting this out. But yes you got it correct. Eviction and pre-fetch overlap when they operate on the same refs. Happy to discuss this more if you have more questions.